pandas/core/groupby.py

```diff
@@ -1574,19 +1641,27 @@ def aggregate(self, values, how, axis=0):
         else:
             out_dtype = 'object'
 
-        # will be filled in Cython function
-        result = np.empty(out_shape, dtype=out_dtype)
-        result.fill(np.nan)
-        counts = np.zeros(self.ngroups, dtype=np.int64)
 
-        result = self._aggregate(result, counts, values, agg_func, is_numeric)
+        if kind == 'aggregate':
+            result = np.empty(out_shape, dtype=out_dtype)
```

Comment:
can move the result outside of this. (e.g. first 2 lines).

use an assert at the very top of the function (that its aggregate/transform), these are internal state